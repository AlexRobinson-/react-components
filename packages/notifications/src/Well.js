import PropTypes from 'prop-types';
import styled from 'styled-components';
import classNames from 'classnames';
import CalloutStyles from '@zendesk/garden-css-callouts/dist/index.css';
import { retrieveTheme, isRtl } from 'garden-react-theming';

const Well = styled.div.attrs({
  className: props => classNames(CalloutStyles['c-callout'], {
    // RTL
    [CalloutStyles['is-rtl']]: isRtl(props),

    // Styles
    [CalloutStyles['c-callout--recessed']]: props.recessed,

    // Validation types
    [CalloutStyles['c-callout--success']]: props.success,
    [CalloutStyles['c-callout--warning']]: props.warning,
    [CalloutStyles['c-callout--error']]: props.error,
  })
})`
  ${props => retrieveTheme('notifications.well', props)}
`;

Well.propTypes = {
  recessed: PropTypes.bool,
  success: PropTypes.bool,
  warning: PropTypes.bool,
  error: PropTypes.bool
};

/** @component */
export default Well;
