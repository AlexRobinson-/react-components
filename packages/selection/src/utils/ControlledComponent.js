import { PureComponent } from 'react';
import PropTypes from 'prop-types';

/** @component */
export default class ControlledComponent extends PureComponent {
  static propTypes = {
    onStateChange: PropTypes.func
  };

  isControlledProp = (key) => {
    return this.props.hasOwnProperty(key)
  }

  /**
   * Used to help retrieve state that can be controlled through props
   */
  getControlledState = (stateToMerge = this.state) => {
    return Object.keys(stateToMerge).reduce((state, key) => {
      state[key] = this.isControlledProp(key)
        ? this.props[key]
        : stateToMerge[key]
      return state
    }, {});
  };

  /**
   * Used to help set state that can be controlled through props
   */
  setControlledState = (newState = {}) => {
    const { onStateChange } = this.props;
    const currentState = this.getControlledState();

    if (onStateChange) {
      const newInternalState = {};
      Object.keys(newState).forEach(key => {
        if (currentState[key] !== newState[key]) {
          newInternalState[key] = newState[key];
        }
      });

      onStateChange(newInternalState);
    } else {
      this.setState(newState);
    }
  };
}
