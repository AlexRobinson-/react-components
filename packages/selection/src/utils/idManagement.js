let idCounter = 1;

/**
 * This generates a unique ID for all autocomplete inputs
 * @param {String} prefix the prefix for the id
 * @return {String} the unique ID
 *
 * @component
 */
function generateId(prefix = 'garden') {
  return `${prefix}-${idCounter++}`
}

/**
 * This is only used in tests... Could be useful in SSR?
 * @param {Number} num The number to set the idCountry to
 *
 * @component
 */
function setIdCounter(num) {
  idCounter = num
}

export default {
  generateId,
  setIdCounter
};
