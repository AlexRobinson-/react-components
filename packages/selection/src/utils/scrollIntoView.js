
function findParent(finder, node, rootNode) {
  if (node !== null && node !== rootNode.parentNode) {
    if (finder(node)) {
      if (node === document.body && node.scrollTop === 0) {
        // in chrome body.scrollTop always return 0
        return document.documentElement
      }
      return node
    } else {
      return findParent(finder, node.parentNode, rootNode)
    }
  } else {
    return null
  }
}

/**
 * Get the closest element that scrolls
 * @param {HTMLElement} node - the child element to start searching for scroll parent at
 * @param {HTMLElement} rootNode - the root element of the component
 * @return {HTMLElement} the closest parentNode that scrolls
 */
const getClosestScrollParent = findParent.bind(
  null,
  node => node.scrollHeight > node.clientHeight + 4,
)

/**
 * Scroll node into view if necessary
 * @param {HTMLElement} node - the element that should scroll into view
 * @param {HTMLElement} rootNode - the root element of the component
 */
// eslint-disable-next-line complexity
export default function scrollIntoView(node, rootNode) {
  const scrollParent = getClosestScrollParent(node, rootNode)
  if (scrollParent === null) {
    return
  }
  const scrollParentStyles = getComputedStyle(scrollParent)
  const scrollParentRect = scrollParent.getBoundingClientRect()
  const scrollParentBorderTopWidth = parseInt(
    scrollParentStyles.borderTopWidth,
    10,
  )
  const scrollParentBorderBottomWidth = parseInt(
    scrollParentStyles.borderBottomWidth,
    10,
  )
  const bordersWidth =
    scrollParentBorderTopWidth + scrollParentBorderBottomWidth
  const scrollParentTop = scrollParentRect.top + scrollParentBorderTopWidth
  const nodeRect = node.getBoundingClientRect()

  if (nodeRect.top < 0 && scrollParentRect.top < 0) {
    scrollParent.scrollTop += nodeRect.top
    return
  }

  if (nodeRect.top < 0) {
    // the item is above the viewport and the parent is not above the viewport
    scrollParent.scrollTop += nodeRect.top - scrollParentTop
    return
  }

  if (nodeRect.top > 0 && scrollParentRect.top < 0) {
    if (
      scrollParentRect.bottom > 0 &&
      nodeRect.bottom + bordersWidth > scrollParentRect.bottom
    ) {
      // the item is below scrollable area
      scrollParent.scrollTop +=
        nodeRect.bottom - scrollParentRect.bottom + bordersWidth
    }
    // item and parent top are on different sides of view top border (do nothing)
    return
  }

  const nodeOffsetTop = nodeRect.top + scrollParent.scrollTop
  const nodeTop = nodeOffsetTop - scrollParentTop
  if (nodeTop < scrollParent.scrollTop) {
    // the item is above the scrollable area
    scrollParent.scrollTop = nodeTop
  } else if (
    nodeTop + nodeRect.height + bordersWidth >
    scrollParent.scrollTop + scrollParentRect.height
  ) {
    // the item is below the scrollable area
    scrollParent.scrollTop =
      nodeTop + nodeRect.height - scrollParentRect.height + bordersWidth
  }
  // the item is within the scrollable area (do nothing)
}
