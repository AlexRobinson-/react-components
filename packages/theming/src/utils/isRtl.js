/** @component */
export default function isRtl(props) {
  return props.theme && props.theme.rtl;
}
