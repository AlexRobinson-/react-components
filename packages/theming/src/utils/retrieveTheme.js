/** @component */
export default function retrieveTheme(componentId, props) {
  const styles = props.theme.styles;

  if (!styles) {
    return;
  }

  const componentStyles = styles[componentId];
  if (typeof componentStyles === 'function') {
    return componentStyles(props);
  }

  return componentStyles;
}
