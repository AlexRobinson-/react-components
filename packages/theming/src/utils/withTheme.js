import { withTheme as styledWithTheme } from 'styled-components';

/** @component */
export default function withTheme(WrappedComonent) {
  return styledWithTheme(WrappedComonent);
};
