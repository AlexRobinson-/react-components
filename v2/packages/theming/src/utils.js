import { withTheme } from 'styled-components';

function retrieveTheme(componentId, props) {
    const styles = props.theme.styles;

    if (!styles) {
        return;
    }

    const componentStyles = styles[componentId];
    if (typeof componentStyles === 'function') {
        return componentStyles(props);
    }

    return componentStyles;
}

function isRtl(props) {
    return props.theme && props.theme.rtl;
}

export default {
    isRtl,
    withTheme,
    retrieveTheme
};
