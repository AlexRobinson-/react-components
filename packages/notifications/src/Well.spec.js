import React from 'react';
import { shallow } from 'enzyme';
import { shallowWithTheme } from 'utils';

import Well from './Well';

it('should render default well styling', () => {
  const wrapper = shallow(<Well />)
  expect(wrapper).toMatchSnapshot();
});

it('should apply RTL styling styling if applied', () => {
  const wrapper = shallowWithTheme(<Well />, true);
  expect(wrapper).toMatchSnapshot();
});

describe('validation', () => {
  it('should render success styling correctly', () => {
    const wrapper = shallow(<Well success />)
    expect(wrapper).toMatchSnapshot();
  });

  it('should render warning styling correctly', () => {
    const wrapper = shallow(<Well warning />)
    expect(wrapper).toMatchSnapshot();
  });

  it('should render error styling correctly', () => {
    const wrapper = shallow(<Well error />)
    expect(wrapper).toMatchSnapshot();
  });
});
