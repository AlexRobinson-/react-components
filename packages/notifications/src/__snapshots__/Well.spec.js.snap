// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should apply RTL styling styling if applied 1`] = `
<div
  className="Well-kOxQyt c-callout is-rtl ikqmIv"
/>
`;

exports[`should render default well styling 1`] = `
<div
  className="Well-kOxQyt c-callout ikqmIv"
/>
`;

exports[`validation should render error styling correctly 1`] = `
<div
  className="Well-kOxQyt c-callout c-callout--error ikqmIv"
/>
`;

exports[`validation should render success styling correctly 1`] = `
<div
  className="Well-kOxQyt c-callout c-callout--success ikqmIv"
/>
`;

exports[`validation should render warning styling correctly 1`] = `
<div
  className="Well-kOxQyt c-callout c-callout--warning ikqmIv"
/>
`;
